Overhaul kim_property docs (#1)
yafshar/kim_commands: kim_commands.rst

* index\_range = KIM portable model parameter index range (an integer for a single element, or pair of integers separated by a colon for a range of elements)
* values = new value(s) to replace the current value(s) of a KIM portable model parameter
* instance\_id = a positive integer identifying the KIM property instance

> * property\_id = name of a `KIM property definition <https://openkim.org/properties>`_ or the full, unique ID of the property including the contributor and date

Reviewer: ```suggestion
* property\_id = name of a `KIM property definition <https://openkim.org/properties>`_, the full, unique ID of the property (including the contributor and date), or a file name corresponding to a local property definition file
```

> * key\_name = one of the keys belonging to the specified KIM property definition (defined by the author of the property definition)

Reviewer: ```suggestion
* key\_name = one of the keys belonging to the specified KIM property definition
```

.. _kim\_property command:

> Converting material properties computed in LAMMPS to KIM property instances (*kim\_property*)

Reviewer: ```suggestion
Writing material properties computed in LAMMPS to stadard KIM property instance format (*kim\_property*)
```

^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

.. note::

   The `kim-property Python package <https://github.com/openkim/kim-property>`_
   requires Python 3.6 or later.

> 

Reviewer: Remove; this is given below in Requirements.

The OpenKIM system includes a collection of tests, models, predictions, and
reference data. A test can be a LAMMPS script that couples with an
interatomic potential to generate one or more predictions, each of which is
associated with a specific material property, and every material property is
associated with a property definition (see the
`KIM Properties Framework <https://openkim.org/doc/schema/properties-framework/>`_
for further details). A prediction of a material property is thus a specific
numerical realization of a property definition, referred to as a "property
instance."  The objective of the *kim\_property* command is to make it as easy
as possible to convert a LAMMPS script that computes a material property into a

> KIM test.

Reviewer: ```suggestion
instance."  The objective of the *kim\_property* command is to make it easy to
output material properties in a standardized, machine readable, format that can be easily ingested by other programs.  Additionally, it aims to make it as easy as possible to convert a LAMMPS script that computes a
material property into a KIM test.
```

   kim_property remove  instance_id key key_name
   kim_property dump    file

Here, *instance\_id* is a positive integer used to uniquely identify each
property instance and, as such, these values cannot be repeated across multiple
property instances.  A *property\_id* is any of (1) a KIM property name, (2)
the full ID of the property including the contributor and date, or (3) a file
name that contains a local user-defined property definition.  Examples of each
of these cases are shown below:

.. parsed-literal::

   kim_property create 1 atomic-mass
   kim_property create 2 cohesive-energy-relation-cubic-crystal

.. parsed-literal::

   kim_property create 1 tag:brunnels@noreply.openkim.org,2016-05-11:property/atomic-mass
   kim_property create 2 tag:staff@noreply.openkim.org,2014-04-15:property/cohesive-energy-relation-cubic-crystal

.. parsed-literal::

   kim_property create 1 new-property.edn

>    kim_property create 2 cohesive-energy-relation-cubic-crystal

Reviewer: ```suggestion
   kim_property create 2 /home/mary/marys-kim-properties/dissociation-energy.edn
```

In the last example, "new-property.edn" is the name of a file that contains a

> user-defined (local) property definition.

Reviewer: ```suggestion
In the last example, "new-property.edn" and "/home/mary/marys-kim-properties/dissociation-energy.edn" are the
names of files that contain user-defined (local) property definitions.
```

A KIM property instance takes the form of a "map," i.e. a set of key-value
pairs akin to Perl\'s hash, Python\'s dictionary, or Java\'s Hashtable.  It
consists of a set of property key names, each of which is referred to here by
the *key\_name* argument, that are defined as part of the relevant property
definition and include only lowercase alphanumeric characters and dashes.  The
value paired with each property key is itself a map whose possible keys are
defined as part of the `KIM Properties Framework
<https://openkim.org/doc/schema/properties-framework>`_; these keys are
referred to by the *key\_name\_key* argument and their associated values by the
*key\_name\_value* argument.  These values may either be scalars or arrays,
as stipulated in the property definition.

.. note::

    Each map assigned to a *key\_name* must contain the *key\_name\_key*
    "source-value" and an associated *key\_name\_value* of the appropriate

>     type (as defined in the relevant property definition).  For keys that are

Reviewer 2: @dskarls  "that are" has been repeated twice.

Author: fixed

    defined as having physical units in their property definition, the
    "source-unit" *key\_name\_key* must also be given a string value recognized
    by `GNU units <https://www.gnu.org/software/units/>`_.

Once a *kim\_property create* command has been given to instantiate a property
instance, maps associated with the property's keys can be edited using the
*kim\_property modify* command.  In using this command, the special keyword
"key" should be given, followed by the property key name and the key-value pair

> in the map associated with that key that is to be set.  For example, the

Reviewer: ```suggestion
in the map associated with the key that is to be set.  For example, the
```

`atomic-mass <https://openkim.org/properties/show/2016-05-11/brunnels@noreply.openkim.org/atomic-mass>`_
property definition consists of two property keys named "mass" and "species."
An instance of this property could be created like so:

.. parsed-literal::

   kim_property create 1 atomic-mass
   kim_property modify 1 key species source-value Al
   kim_property modify 1 key mass    source-value 26.98154
   kim_property modify 1 key mass    source-unit amu

or, equivalently,

.. parsed-literal::

   kim_property create 1 atomic-mass

>    kim_property modify 1 key species source-value Al       &
>                          key mass    source-value 26.98154 &
>                                      source-unit  amu

Reviewer: I don't think this can work as expected.  I think it needs to be:
```suggestion
   kim_property modify 1 key species source-value Al       &
                         key mass    source-value 26.98154 &
                         key mass    source-unit  amu
```

Reviewer 2: @ellio167 it is working as expected. We do not need to add the key again.

property definition name, and creates an initial empty property instance data
structure.  For example,

.. parsed-literal::

   kim_property create 1 atomic-mass
   kim_property create 2 cohesive-energy-relation-cubic-crystal

creates an empty property instance of the "atomic-mass" property definition
with instance ID 1 and an empty instance of the
"cohesive-energy-relation-cubic-crystal" property with ID 2.  A list of
published property definitions in OpenKIM can be found on the `properties page
<https://openkim.org/properties>`_.

One can also provide the name of a file in the current working directory or the
path of a file containing a valid property definition.  For example,

.. parsed-literal::

   kim_property create 1 new-property.edn

where "new-property.edn" refers to a file name containing a new property
definition that does not exist in OpenKIM.

If the *property\_id* given cannot be found in OpenKIM and no file of this name
containing a valid property definition can be found, this command will produce
an error with an appropriate message.  Calling *kim\_property create* with the
same instance ID multiple times will also produce an error.

> 

Reviewer: This seems to be mostly repeated.  Remove, but put any new information above where `create` is first discussed.

**Destroy**

.. parsed-literal::

   kim_property destroy instance_id

The *kim\_property destroy* command deletes a previously created property
instance ID.  For example,

.. parsed-literal::

   kim_property destroy 2

Note that if this command is called with an instance ID that does not exist, no
error is raised.

> 

Reviewer: move this below the complete discussion of `modify`

Reviewer 2: @ellio167 do you mean the whole destroy part?

> **Modify**

Reviewer: Merge this with the above discussion of `modify` and remove any repeated content

where the special keyword "key" is followed by a *key\_name* ("species" or
"mass" in the above) and one or more key-value pairs.  These key-value pairs
may continue until either another "key" keyword is given or the end of the
command line is reached.  Thus, the above could equivalently be written as

.. parsed-literal::

   kim_property modify 1 key species source-value Al       &
                         key mass    source-value 26.98154 &
                         key mass    source-unit  amu

As an example of modifying multiple key-value pairs belonging to the map of a
single property key, the following command modifies the map of the
"cohesive-potential-energy" property key to contain the key "source-unit" which
is assigned a value of "eV" and the key "digits" which is assigned a value of
5:

.. parsed-literal::

   kim_property modify 2 key cohesive-potential-energy source-unit eV digits 5

Note that the relevant data types of the values in the map are handled
automatically based on the `KIM Properties Framework
<https://openkim.org/doc/schema/properties-framework>`_.  In the example above,
this means that the value "eV" will automatically be interpreted as a string

> while the value 5 will be interpreted as an integer.

Reviewer: again, I don't think this works as expected.

Reviewer 2: @ellio167 would you please explain why do you think that?

Reviewer: Maybe it is OK.  I was thinking that the command would not know how to correctly parse this because there could be a variable number of arguments after the 'key' argument.  However, I guess that is the old way we had things done...

I guess, now the number of arguments is just one, unless a range is given (e.g., 2:5).  So, yes, it seems that this should be OK.

If you agree, then I'm good with this as is.

   be assigned default values based on the data type associated with the key in
   the map:

   +-----------+---------------+
   | Data type | Default value |
   +-----------+---------------+
   | int       |  0            |
   +-----------+---------------+
   | float     |  0.0          |
   +-----------+---------------+
   | bool      |  False        |
   +-----------+---------------+
   | string    |  ""           |
   +-----------+---------------+
   | file      |  ""           |
   +-----------+---------------+

   For example, doing the following:

   .. parsed-literal::

      kim_property create 2 cohesive-energy-relation-cubic-crystal
      kim_property modify 2 key species source-value 4 Al

   will result in the "source-value" key in the map for the property key
   "species" being assigned the array ["", "", "", "Al"].

For convenience, the index argument provided may refer to an inclusive range of
indices by specifying two integers separated by a colon (the first integer must
be less than or equal to the second integer, and no whitespace should be
included).  Thus, the snippet above could equivalently be written:

.. parsed-literal::

> 

Reviewer 2: @dskarls You have removed the below example. But I think it is good to have it here. Especially as it shows how to use the property command within a loop.

For example,

```
   label       loopa
   variable    i loop 5
   ...
   kim_property modify 2 key a source-value $i ${a_value}
   ...
   next        i
   jump        SELF loopa
```

Author: I had decided to remove it because there's no logic in the example that's specific to the 'kim_property' command.  If someone has read the 'kim_property' docs and has read the docs that explain loops, they can combine them without knowing anything special.

Reviewer 2: @dskarls it makes sense

    [0.5, 0.0, 0.5],
    [0.0, 0.5, 0.5]]

> While each of the nine components could be set individually, we can instead set

Reviewer 2: @dskarls Here "nine" components should be "twelve"?

Author: fixed

                         key basis-atom-coordinates source-value 3 1:3 0.5 0.0 0.5 &
                         key basis-atom-coordinates source-value 4 1:3 0.0 0.5 0.5

.. parsed-literal::

>     kim_property modify 2 key short-name source-value 1 fcc                         &

Reviewer: Not sure about the `key a` part of this one
